{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Results from *Michael Heinzinger, Konstantin Weissenow, Joaquin Gomez Sanchez, Adrian Henkel, Milot Mirdita, Martin Steinegger, Burkhard Rost, Bilingual language model for protein sequence and structure, NAR Genomics and Bioinformatics, Volume 6, Issue 4, December 2024, lqae150, https://doi.org/10.1093/nargab/lqae150*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**CATH**\n",
    "\n",
    "hierarchical classification scheme for protein domain structures, based on **C**lass, **A**rchitecture, **T**opology (or fold), and **H**omologous superfamily\n",
    "\n",
    "Class (C): Broad structural categorization based on secondary structure composition.\n",
    "Main classes:\n",
    "- Mainly alpha (α-helices dominate)\n",
    "- Mainly beta (β-sheets dominate)\n",
    "- Alpha-beta (mixed α/β structures)\n",
    "- Few secondary structures (little secondary structure)\n",
    "\n",
    "Architecture (A): Arrangements of secondary structures in 3D space, independent of connectivity.\n",
    "\n",
    "Topology (T) / Fold: Describes the connectivity of secondary structures, defining unique folds.\n",
    "\n",
    "Homologous Superfamily (H): Groups protein domains that share a common ancestor and have structural and functional similarities."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from Supplements Table S1 \"ProstT5 embeddings for fold classification (contrastive learning)\"\n",
    "\n",
    "contrastive_learning_results = {\n",
    "    \"ESM-1b\": [87, 4, 68, 6, 59, 7, 70, 7, 71, 6],\n",
    "    \"Ankh\":\t[87, 4, 76, 6, 64, 6, 75, 7, 75, 6],\n",
    "    \"ProtT5\": [89, 4, 75, 6, 64, 6, 76, 6, 76, 6],\n",
    "    \"ProstT5(AA)\": [89, 4, 81, 5, 70, 6, 83, 6, 81, 5],\n",
    "    \"ProstT5(p3Di)\": [88, 4, 77, 5, 69, 6, 80, 6, 79, 5],\n",
    "    \"ProstT5(3Di)\":\t[90, 4,\t78, 5, 69, 6, 74, 7, 78, 6],\n",
    "    \"ProstT5(cat)\":\t[91, 4, 79, 5, 72, 6, 84, 6, 81, 5] # concatenation of ProstT5(AA) and ProstT5(p3Di)\n",
    "}\n",
    "\n",
    "contrastive_learning_results_column_names = [\"Model\",\"ctl_C\",\"ctl_C_ci\", \"ctl_A\", \"ctl_A_ci\", \"ctl_T\", \"ctl_T_ci\", \"ctl_H\",\"ctl_H_ci\", \"ctl_Mean\", \"ctl_Mean_ci\"]\n",
    "\n",
    "# trun into a pandas dataframe\n",
    "contrastive_learning_results_df = pd.DataFrame(contrastive_learning_results).T\n",
    "# reset the index\n",
    "contrastive_learning_results_df.reset_index(inplace=True)\n",
    "# rename the columns\n",
    "contrastive_learning_results_df.columns = contrastive_learning_results_column_names\n",
    "contrastive_learning_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from Table 1 \"Classification of proteins into CATH hierarchy (folds)*\"\n",
    "\n",
    "unsupervised_results = {\n",
    "    \"ESM-1b\": [79, 5, 61, 6, 50, 7, 57, 8, 62, 7],\n",
    "    \"Ankh\": [84, 5, 69, 6, 60, 7, 67, 8, 70, 6],\n",
    "    \"ProtT5\":\t[84, 5, 67, 6, 57, 6, 64, 8, 68, 6],\n",
    "    \"ProstT5(AA)\":\t[85, 5, 74, 6, 64, 6, 69, 7, 73, 6],\n",
    "    \"ProstT5(p3Di)\":    [85, 5, 71, 6, 60, 7, 73, 7, 72, 6],\n",
    "    \"ProstT5(3Di)\":\t[90, 4, 77, 6, 65, 6, 75, 7, 77, 6],\n",
    "    \"ProstT5(cat)\":\t[88, 4, 74, 6, 65, 7, 74, 7, 75, 6]\n",
    "}\n",
    "\n",
    "unsupervised_results_column_names = [\"Model\",\"C\",\"C_ci\", \"A\", \"A_ci\", \"T\", \"T_ci\", \"H\",\"H_ci\", \"Mean\", \"Mean_ci\"]\n",
    "\n",
    "# trun into a pandas dataframe\n",
    "unsupervised_results_df = pd.DataFrame(unsupervised_results).T\n",
    "# reset the index\n",
    "unsupervised_results_df.reset_index(inplace=True)\n",
    "# rename the columns\n",
    "unsupervised_results_df.columns = unsupervised_results_column_names\n",
    "unsupervised_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine the two dataframes\n",
    "combined_results_df = pd.concat([contrastive_learning_results_df, unsupervised_results_df], axis=1)\n",
    "# drop the duplicate column\n",
    "combined_results_df = combined_results_df.loc[:,~combined_results_df.columns.duplicated()]\n",
    "combined_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the results as scatter plot compraing clt to unsupervised\n",
    "\n",
    "fig = px.scatter(combined_results_df, x=\"Mean\", y=\"ctl_Mean\", text=\"Model\")\n",
    "fig.update_traces(textposition='top center')\n",
    "fig.update_layout(title=\"Protein Fold Classification Accuracy (Mean)\",\n",
    "                  template=\"plotly_white\",\n",
    "                  font=dict(color=\"black\", family=\"Arial\"),\n",
    "                  width=500,\n",
    "                  height=500,\n",
    "                  xaxis_title=\"Unsupervised Learning\",\n",
    "                  yaxis_title=\"Contrastive Learning\")\n",
    "# adapt axis range\n",
    "fig.update_xaxes(range=[60, 85])\n",
    "fig.update_yaxes(range=[60, 85])\n",
    "fig.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the results as scatter plot compraing clt to unsupervised\n",
    "\n",
    "fig = px.scatter(combined_results_df, x=\"Mean\", y=\"ctl_Mean\", text=\"Model\",\n",
    "                 # add the error bars\n",
    "                 error_x=\"Mean_ci\", error_y=\"ctl_Mean_ci\",)\n",
    "fig.update_traces(textposition='top center',\n",
    "                  error_x=dict(color=\"lightgrey\"),\n",
    "                  error_y=dict(color=\"lightgrey\") )\n",
    "fig.update_layout(title=\"Protein Fold Classification Accuracy (Mean)\",\n",
    "                  template=\"plotly_white\",\n",
    "                  font=dict(color=\"black\", family=\"Arial\"),\n",
    "                  width=500,\n",
    "                  height=500,\n",
    "                  xaxis_title=\"Unsupervised Learning\",\n",
    "                  yaxis_title=\"Contrastive Learning\")\n",
    "# adapt the range of the x and y axis\n",
    "fig.update_xaxes(range=[60, 85])\n",
    "fig.update_yaxes(range=[60, 85])\n",
    "fig.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the results as scatter plot compraing clt to unsupervised\n",
    "\n",
    "fig = px.scatter(combined_results_df, x=\"A\", y=\"ctl_A\", text=\"Model\",\n",
    "                 # add the error bars\n",
    "                 error_x=\"A_ci\", error_y=\"ctl_A_ci\",)\n",
    "fig.update_traces(textposition='top center',\n",
    "                  error_x=dict(color=\"lightgrey\"),\n",
    "                  error_y=dict(color=\"lightgrey\") )\n",
    "fig.update_layout(title=\"Protein Fold Classification Accuracy (CATH-A)\",\n",
    "                  template=\"plotly_white\",\n",
    "                  font=dict(color=\"black\", family=\"Arial\"),\n",
    "                  width=500,\n",
    "                  height=500,\n",
    "                  xaxis_title=\"Unsupervised Learning\",\n",
    "                  yaxis_title=\"Contrastive Learning\")\n",
    "# adapt the range of the x and y axis\n",
    "fig.update_xaxes(range=[50, 85])\n",
    "fig.update_yaxes(range=[50, 85])\n",
    "fig.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the results as scatter plot compraing clt to unsupervised\n",
    "\n",
    "fig = px.scatter(combined_results_df, x=\"C\", y=\"ctl_C\", text=\"Model\",\n",
    "                 # add the error bars\n",
    "                 error_x=\"C_ci\", error_y=\"ctl_C_ci\",)\n",
    "fig.update_traces(textposition='top center',\n",
    "                  error_x=dict(color=\"lightgrey\"),\n",
    "                  error_y=dict(color=\"lightgrey\") )\n",
    "fig.update_layout(title=\"Protein Fold Classification Accuracy (CATH-C)\",\n",
    "                  template=\"plotly_white\",\n",
    "                  font=dict(color=\"black\", family=\"Arial\"),\n",
    "                  width=500,\n",
    "                  height=500,\n",
    "                  xaxis_title=\"Unsupervised Learning\",\n",
    "                  yaxis_title=\"Contrastive Learning\")\n",
    "# adapt the range of the x and y axis\n",
    "fig.update_xaxes(range=[70, 100])\n",
    "fig.update_yaxes(range=[70, 100])\n",
    "fig.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the results as scatter plot compraing clt to unsupervised\n",
    "\n",
    "fig = px.scatter(combined_results_df, x=\"T\", y=\"ctl_T\", text=\"Model\",\n",
    "                 # add the error bars\n",
    "                 error_x=\"T_ci\", error_y=\"ctl_T_ci\",)\n",
    "fig.update_traces(textposition='top center',\n",
    "                  error_x=dict(color=\"lightgrey\"),\n",
    "                  error_y=dict(color=\"lightgrey\") )\n",
    "fig.update_layout(title=\"Protein Fold Classification Accuracy (CATH-T)\",\n",
    "                  template=\"plotly_white\",\n",
    "                  font=dict(color=\"black\", family=\"Arial\"),\n",
    "                  width=500,\n",
    "                  height=500,\n",
    "                  xaxis_title=\"Unsupervised Learning\",\n",
    "                  yaxis_title=\"Contrastive Learning\")\n",
    "# adapt the range of the x and y axis\n",
    "fig.update_xaxes(range=[40, 80])\n",
    "fig.update_yaxes(range=[40, 80])\n",
    "fig.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the results as scatter plot compraing clt to unsupervised\n",
    "\n",
    "fig = px.scatter(combined_results_df, x=\"H\", y=\"ctl_H\", text=\"Model\",\n",
    "                 # add the error bars\n",
    "                 error_x=\"H_ci\", error_y=\"ctl_H_ci\",)\n",
    "fig.update_traces(textposition='top center',\n",
    "                  error_x=dict(color=\"lightgrey\"),\n",
    "                  error_y=dict(color=\"lightgrey\") )\n",
    "fig.update_layout(title=\"Protein Fold Classification Accuracy (CATH-H)\",\n",
    "                  template=\"plotly_white\",\n",
    "                  font=dict(color=\"black\", family=\"Arial\"),\n",
    "                  width=500,\n",
    "                  height=500,\n",
    "                  xaxis_title=\"Unsupervised Learning\",\n",
    "                  yaxis_title=\"Contrastive Learning\")\n",
    "# adapt the range of the x and y axis\n",
    "fig.update_xaxes(range=[50, 90])\n",
    "fig.update_yaxes(range=[50, 90])\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
